src/dscanner/analysis/unused_result.d

```diff
@@ -24,174 +16,186 @@ import std.range : retro;
-        if (!ue)
-            return;
-        auto fce = ue.functionCallExpression;
-        if (!fce)
-            return;
-
-        auto identVisitor = scoped!IdentVisitor;
-        if (fce.unaryExpression !is null)
-            identVisitor.visit(fce.unaryExpression);
-        else if (fce.type !is null)
-            identVisitor.visit(fce.type);
-
-        if (!identVisitor.names.length)
-            return;
-
-        const(DSymbol)*[] symbols = resolveSymbol(sc, identVisitor.names);
-
-        if (!symbols.length)
-            return;
-
-        foreach (sym; symbols)
-        {
-            if (!sym)
-                return;
-            if (!sym.type)
-                return;
-            if (sym.kind != CompletionKind.functionName)
-                return;
-            if (sym.type is void_)
-                return;
-            if (noreturn_ && sym.type is noreturn_)
-                return;
-        }
-
-        const(Token)[] tokens = fce.unaryExpression
-            ? fce.unaryExpression.tokens
-            : fce.type
-            ? fce.type.tokens
-            : fce.tokens;
-
-        addErrorMessage(tokens, KEY, MSG);
-    }
+	alias visit = BaseAnalyzerDmd.visit;
+	mixin AnalyzerInfo!"unused_result";
+	private enum KEY = "dscanner.performance.enum_array_literal";
+	private enum string MSG = "Function return value is discarded";
+
+	extern (D) this(string fileName, bool skipTests = false)
+	{
+		super(fileName, skipTests);
+	}
+
+	mixin VisitInstructionBlock!(AST.WhileStatement);
+	mixin VisitInstructionBlock!(AST.ForStatement);
+	mixin VisitInstructionBlock!(AST.DoStatement);
+	mixin VisitInstructionBlock!(AST.ForeachRangeStatement);
+	mixin VisitInstructionBlock!(AST.ForeachStatement);
+	mixin VisitInstructionBlock!(AST.SwitchStatement);
+	mixin VisitInstructionBlock!(AST.SynchronizedStatement);
+	mixin VisitInstructionBlock!(AST.WithStatement);
+	mixin VisitInstructionBlock!(AST.TryCatchStatement);
+	mixin VisitInstructionBlock!(AST.TryFinallyStatement);
+
+	override void visit(AST.CompoundStatement compoundStatement)
+	{
+		foreach (statement; *compoundStatement.statements)
+		{
+			if (hasUnusedResult(statement))
+			{
+				auto lineNum = cast(ulong) statement.loc.linnum;
+				auto charNum = cast(ulong) statement.loc.charnum;
+				addErrorMessage(lineNum, charNum, KEY, MSG);
+			}
+
+			statement.accept(this);
+		}
+	}
+
+	override void visit(AST.IfStatement ifStatement)
+	{
+		if (hasUnusedResult(ifStatement.ifbody))
+		{
+			auto lineNum = cast(ulong) ifStatement.ifbody.loc.linnum;
+			auto charNum = cast(ulong) ifStatement.ifbody.loc.charnum;
+			addErrorMessage(lineNum, charNum, KEY, MSG);
+		}
+
+		if (ifStatement.elsebody && hasUnusedResult(ifStatement.elsebody))
+		{
+			auto lineNum = cast(ulong) ifStatement.elsebody.loc.linnum;
+			auto charNum = cast(ulong) ifStatement.elsebody.loc.charnum;
+			addErrorMessage(lineNum, charNum, KEY, MSG);
+		}
+
+		super.visit(ifStatement);
+	}
```

Comment:
I don't understand why you need separate visitation methods for ifs and compound statements. This is more elegantly solved by simply transforming the `hasUnusedResult` function into a generic statement visitation function that simply calls `super.visit()` on the statement parameter at the end of the function.

Reply:
Ah nvm that. I understand now.